beacon-chain/db/kv/state.go

```diff
@@ -318,3 +320,44 @@ func createStateIndicesFromStateSlot(ctx context.Context, slot uint64) map[strin
 	}
 	return indicesByBucket
 }
+
+// CleanUpDirtyStates removes states in DB that falls to under archived point interval rules.
+// Only following states would be kept:
+// 1.) state_slot % archived_interval == 0. (e.g. archived_interval=2048, states with slot 2048, 4096... etc)
+// 2.) archived_interval - slots_per_epoch/2 < state_slot % archived_interval
+//   (e.g. archived_interval=2048, slots_per_epoch=32, states with slot 2047, 2046, 2032... etc).
+//   This is to tolerate skip slots. Not every state lays on the boundary.
+// 3.) state with current finalized root
+func (s *Store) CleanUpDirtyStates(ctx context.Context, slotsPerArchivedPoint uint64) error {
+	ctx, span := trace.StartSpan(ctx, "BeaconDB.CleanUp")
+	defer span.End()
+	f, err := s.FinalizedCheckpoint(ctx)
+	if err != nil {
+		return err
+	}
+	deletedRoots := make([][32]byte, 0)
+
+	err = s.db.View(func(tx *bolt.Tx) error {
+		bkt := tx.Bucket(stateSlotIndicesBucket)
+		if err := bkt.ForEach(func(k, v []byte) error {
+			finalized := bytesutil.ToBytes32(f.Root) == bytesutil.ToBytes32(v)
+			slot := bytesutil.BytesToUint64BigEndian(k)
+			mod := slot % slotsPerArchivedPoint
+			// The following conditions cover 1, 2, and 3 above.
+			if mod != 0 && mod < slotsPerArchivedPoint-params.BeaconConfig().SlotsPerEpoch/2 && !finalized {
```

Comment:
Shouldn't we use non-strict inequality `mod <= slotsPerArchivedPoint-params.BeaconConfig().SlotsPerEpoch/2`?

Reply:
Yeah I think you are right. I'll update test to reflect that as well